File: julia/wired.jl
""" Wired.jl - Julia Benchmark 
    (c) 2024 ryan@freestatelabs.com 
"""

using LinearAlgebra: norm, dot
const mu0 = 4pi * (1e-7)

# Create the benchmark problem with `N` nodes and `M` sources in a row-major format
#  (slower). `nodes` is 3xN, `sources` is 7xM. 
function createproblem_rm(N::Integer, M::Integer; I=5000, Px=3, Py=3)
    nodes = zeros(3,N) 
    sources = zeros(7,M) 

    zstart = 0; zend = 1
    dz = (zend - zstart) / (N-1)
    z = zstart

    for i = 1:N
        nodes[:,i] = [Px, Py, z]
        z += dz 
    end

    zstart = -1000; zend = 1000 
    dz = (zend - zstart) / M
    z = zstart 
    for i = 1:M 
        sources[:,i] = [0, 0, z, 0, 0, z+dz, I] 
        z += dz
    end

    return nodes, sources
end


# Create the benchmark problem with `N` nodes and `M` sources in a column-major 
#   format (faster). Nodes is Nx3, sources is 7xM.
function createproblem(N::Integer, M::Integer; I=5000, Px=3, Py=3)
    nodes = zeros(N,3) 
    sources = zeros(7,M) 

    zstart = 0; zend = 1
    dz = (zend - zstart) / (N-1)
    z = zstart

    for i = 1:N
        nodes[i,:] = [Px, Py, z]
        z += dz 
    end

    zstart = -1000; zend = 1000 
    dz = (zend - zstart) / M
    z = zstart 
    for i = 1:M 
        sources[:,i] = [0, 0, z, 0, 0, z+dz, I] 
        z += dz
    end

    return nodes, sources
end


# In-place cross-product to reduce memory allocations, c = a x b 
# This version is defined for 3-length vectors
# See: https://discourse.julialang.org/t/excessive-memory-allocation/108841/4 
function cross!(c::Vector{Float64}, a::Vector{Float64}, b::Vector{Float64})
    if !(length(a) == length(b) == 3)
        throw(DimensionMismatch("Vectors must be of length=3!"))
    end

    c .= (a[2]*b[3]-a[3]*b[2], a[3]*b[1]-a[1]*b[3], a[1]*b[2]-a[2]*b[1])
end


# In-place cross-product for row-major format; cross the rows of A with b and place in C
@views function cross!(C::AbstractArray{Float64}, A::AbstractArray{Float64}, b::Vector{Float64})

    C[1,:] .= A[2,:] .* b[3] .- A[3,:] .* b[2]
    C[2,:] .= A[3,:] .* b[1] .- A[1,:] .* b[3]
    C[3,:] .= A[1,:] .* b[2] .- A[2,:] .* b[1]

end


# In-place cross-product for column-major format; cross the columns of A with b and place in C
@views function crosscols!(C::AbstractArray{Float64}, A::AbstractArray{Float64}, b::Vector{Float64})

    C[:,1] .= A[:,2] .* b[3] .- A[:,3] .* b[2]
    C[:,2] .= A[:,3] .* b[1] .- A[:,1] .* b[3]
    C[:,3] .= A[:,1] .* b[2] .- A[:,2] .* b[1]

end


# In-place dot-product of two arrays; note that a and b must both be 3xN
@views function dotcols!(c::Vector{Float64}, A::AbstractArray{Float64}, B::AbstractArray{Float64})

    c .= (A[1,:] .* B[1,:] .+ A[2,:] .* B[2,:] .+ A[3,:] .* B[3,:])
end


# In-place dot-product of the rows of A by vector b; place in C
@views function dotrows!(C::Vector{Float64}, A::AbstractArray{Float64}, b::Vector{Float64})

    C .= (A[:,1] .* b[1] .+ A[:,2] .* b[2] .+ A[:,3] .* b[3])
end


# In-place vector norm of the 3-length columns of an array 
#   This is for row-major format (slower)
@views function normcols!(c::Vector{Float64}, A::AbstractArray{Float64})

    c .= (sqrt.(A[1,:].^2 .+ A[2,:].^2 .+ A[3,:].^2))

end


# In-place vector norm of the 3-length rows of an array 
#   This is for column-major format (faster)
@views function normrows!(c::Vector{Float64}, A::AbstractArray{Float64})

    c .= (sqrt.(A[:,1].^2 .+ A[:,2].^2 .+ A[:,3].^2))

end


# Multiply the rows of a matrix `A` by a vector b 
@views function mult!(A::AbstractArray{Float64}, b::Vector{Float64})

    A[1,:] .*= b 
    A[2,:] .*= b 
    A[3,:] .*= b

end


# Multiply the rows of a matrix `A` by a vector b 
@views function multcols!(A::AbstractArray{Float64}, b::Vector{Float64})

    A[:,1] .*= b 
    A[:,2] .*= b 
    A[:,3] .*= b

end


# Solve the Biot-Savart problem with optimization level 1: simple loops
#   This is for row-major implementation; it turns out this is about as fast 
#   as the column-major implementation
@views function solve_rm(nodes::AbstractArray{Float64}, sources::AbstractArray{Float64})

    B = zeros(size(nodes))
    a = Vector{Float64}(undef,3)
    b = similar(a)
    c = similar(a) 
    d = 0.0
    cxa = similar(a)

    for j in axes(sources)[2]
        d = mu0 * sources[7,j] / (4pi)
        a .= sources[4:6,j] .- sources[1:3,j] 
        for i in axes(nodes)[2]
            b .= sources[1:3,j] .- nodes[:,i] 
            c .= sources[4:6,j] .- nodes[:,i]
            cross!(cxa, c, a)

            B[:,i] .+= cxa .* d .* (1/(norm(cxa)^2)) .* (dot(a,c)/norm(c) .- dot(a,b)/norm(b))
        end
    end

    return B
end


# Solve the Biot-Savart problem with optimization level 1: simple loops
@views function solve(nodes::AbstractArray{Float64}, sources::AbstractArray{Float64})

    B = zeros(size(nodes))
    a = Vector{Float64}(undef,3)
    b = similar(a)
    c = similar(a) 
    d = 0.0
    cxa = similar(a)

    for j in axes(sources)[2]
        d = mu0 * sources[7,j] / (4pi)
        a .= sources[4:6,j] .- sources[1:3,j] 
        for i in axes(nodes)[1]
            b .= sources[1:3,j] .- nodes[i,:] 
            c .= sources[4:6,j] .- nodes[i,:]
            cross!(cxa, c, a)

            B[i,:] .+= cxa .* d .* (1/(norm(cxa)^2)) .* (dot(a,c)/norm(c) .- dot(a,b)/norm(b))
        end
    end

    return B
end


# Solve the Biot-Savart problem with optimization level 1: vectorization
# Replace the inner loop with matrix/vector operations to allow Julia to make 
#   low-level optimizations not otherwise visible to the programmer
# This is the original row-major implementation (slower)
@views function solve2_rm(nodes::AbstractArray{Float64}, sources::AbstractArray{Float64})

    Nn = size(nodes)[2]
    B = zeros(3, Nn)
    a = zeros(3) 
    b = zeros(3, Nn) 
    c = zeros(3,Nn) 
    cxa = zeros(3,Nn) 
    norm_cxa = zeros(Nn) 
    dot_ac = zeros(Nn) 
    norm_c = zeros(Nn) 
    dot_ab = zeros(Nn) 
    norm_b = zeros(Nn) 

    d = 0.0

    for j in axes(sources)[2]
        d = mu0 * sources[7,j] / (4pi)
        a .= sources[4:6,j] .- sources[1:3,j] 
        b .= sources[1:3,j] .- nodes 
        c .= sources[4:6,j] .- nodes

        cross!(cxa, c, a)
        normcols!(norm_cxa, cxa)
        dotcols!(dot_ac, a, c) 
        dotcols!(dot_ab, a, b) 
        normcols!(norm_c, c) 
        normcols!(norm_b, b)

        mult!(cxa, (d .* (norm_cxa.^(-2)) .* (dot_ac./norm_c .- dot_ab./norm_b))) 
        B .+= cxa
    end

    return B
end


# Solve the Biot-Savart problem with optimization level 2: vectorization
# Replace the inner loop with matrix/vector operations to allow Julia to make 
#   low-level optimizations not otherwise visible to the programmer
# This is the column-major version (faster)
@views function solve2(nodes::AbstractArray{Float64}, sources::AbstractArray{Float64})

    Nn = size(nodes)[1]
    B = zeros(Nn,3)
    a = zeros(3) 
    b = zeros(Nn,3) 
    c = zeros(Nn,3) 
    cxa = zeros(Nn,3) 
    norm_cxa = zeros(Nn) 
    dot_ac = zeros(Nn) 
    norm_c = zeros(Nn) 
    dot_ab = zeros(Nn) 
    norm_b = zeros(Nn) 

    d = 0.0

    for j in axes(sources)[2]

        # The use of "dot assignment" i.e. `.=` prevents re-allocation of memory 
        #  within the loop
        d = mu0 * sources[7,j] / (4pi)
        a .= sources[4:6,j] .- sources[1:3,j] 
        b .= sources[1:3,j]' .- nodes 
        c .= sources[4:6,j]' .- nodes

        # In-place operators also prevent re-allocation of memory
        crosscols!(cxa, c, a)
        normrows!(norm_cxa, cxa)
        dotrows!(dot_ac, c, a) 
        dotrows!(dot_ab, b, a) 
        normrows!(norm_c, c) 
        normrows!(norm_b, b)

        multcols!(cxa, (d .* (norm_cxa.^(-2)) .* (dot_ac./norm_c .- dot_ab./norm_b))) 
        B .+= cxa
    end

    return B
end


# Split a problem up for multi-threaded operation 
# For thread number `it`, total number of threads `Nt`, and total number of 
#   tasks `N`, determine the start/stop index for that particular thread number
function threadindices(it::Integer, Nt::Integer, N::Integer) 

    Nperthread = div(N, Nt)
    remainder = rem(N, Nt) 

    if it == 1
        i1 = 1 
        i2 = i1 + Nperthread - 1 
    elseif it == Nt 
        i2 = N 
        i1 = i2 - Nperthread - remainder + 1 
    else
        i1 = (it-1)*Nperthread + 1
        i2 = i1 + Nperthread - 1 
    end 

    return i1:i2
end


# Solve the Biot-Savart problem with optimization level 3: multi-threading
# Default is to use all threads available to Julia
@views function solve3(nodes::AbstractArray{Float64}, sources::AbstractArray{Float64}; 
                        Nt::Integer = 0)

    Ns = size(sources)[2]
    if Nt == 0 
        Nt = Threads.nthreads()
    end

    # Spawn a new task for each thread by splitting up the source array
    tasks = Vector{Task}(undef, Nt)
    for it = 1:Nt 
        @views tasks[it] = Threads.@spawn solve2(nodes, sources[:,threadindices(it, Nt, Ns)])
    end
    
    # Get the result from each calculation and add it to the output array 
    B = zeros(size(nodes))
    for it = 1:Nt 
        B .+= fetch(tasks[it]) 
    end 

    return B
end

File: python/wired.py
""" Wired.py 
    Python implementation of the `Wired.jl` code, for benchmarking. 

    Optimization levels:
    0 :: Pure Python, standard library only (no vectorization or NumPy)
    1 :: NumPy
    2 :: Vectorization with NumPy 
    3 :: Vectorization with NumPy + Numba
    4 :: Parallel processing (on local CPU) via Numba
"""

from typing import Union
from math import sqrt, pi
import numpy as np 
from numba import njit, config, get_num_threads, set_num_threads, prange

Number = Union[int, float, np.number]
norm2 = np.linalg.norm

mu0 = 4*pi*(1e-7)

#
# Create the problem 
#

def create_problem(N: int, M: int, I=5000, Px=3, Py=3): 
    """ Make the sample problem: semi-infinite wire along the z-axis carrying 
        5 kA, output nodes are at x=y=3, along the z-axis. 

        N is the number of output nodes, M is the number input sources. Returns
        (nodes, sources), where nodes and sources are lists of lists.
    """

    # Make nodes
    zstart = 0; zend = 1
    dz = (zend - zstart) / (N - 1)
    z = zstart

    nodes = []
    for i in range(0, N):
        nodes.append([Px,Py,z])
        z += dz 

    # Make sources
    zstart = -1000; zend = 1000 
    dz = (zend - zstart) / N 
    z = zstart

    x = y = 0
    sources = [] 
    for i in range(0, M):
        sources.append([x, y, z, x, y, z+dz, I])
        z += dz 

    return nodes, sources


def create_problem2(N: int, M: int, I=5000, Px=3, Py=3): 
    """ Numpy version of the problem. """

    nodes, sources = create_problem(N, M, I, Px, Py)
    return np.array(nodes), np.array(sources)

def create_problem_rm(N: int, M: int, I=5000, Px=3, Py=3): 
    """ Numpy version of the problem, row-major. """

    nodes, sources = create_problem(N, M, I, Px, Py)
    return np.transpose(np.array(nodes)), np.array(sources)


#
# Optimization level 0 - pure Python 
# 

def cross(a: list, b: list):
    """ Cross-product for non-numpy implementation. """

    c1 = a[1]*b[2] - a[2]*b[1]
    c2 = a[2]*b[0] - a[0]*b[2] 
    c3 = a[0]*b[1] - a[1]*b[0]

    return [c1, c2, c3]


def norm(a) -> Number: 
    """ Vector norm for non-numpy implementation. """

    sumsq = 0 
    for a_ in a:
        sumsq += a_**2 

    return sqrt(sumsq)


def dot(a, b) -> Number: 
    """ Dot product of two vectors, for non-numpy implementation. """
    
    c = 0 
    for i in range(0, len(a)):
        c += a[i]*b[i] 

    return c


def minus(a: list, b:list) -> list:
    """ Subtract two lists from each other, for non-numpy implementation. """

    c = []
    for i in range(0, len(a)): 
        c.append(a[i] - b[i])

    return c


def biotsavart(J: list, K: list, I: Number, P: list) -> list:
    """ Opt level 0: compute the magnetic field caused by a single source at a
        single node. 
    """
    a = minus(K, J) 
    b = minus(J, P)
    c = minus(K, P)

    cxa = cross(c,a) 
    s0 = mu0*I/(4*pi) 
    s1 = norm(cxa)**2
    s2 = dot(a,c) 
    s3 = norm(c)
    s4 = dot(a,b)
    s5 = norm(b)

    bfield = cxa 
    for i in range(0, len(bfield)):
        bfield[i] *= s0 * (1 / s1) * (s2 / s3 - s4/s5)

    return bfield


def solve0(nodes: list, sources: list) -> list: 
    """ Optimization level 0: pure Python
    """

    Bfield = [] 

    for node in nodes: 

        b = [0, 0, 0]

        for source in sources: 
            b_ = biotsavart(source[0:3], source[3:6], source[6], node)

            for i in range(0, len(b)):
                b[i] += b_[i]

        Bfield.append(b)

    return Bfield


# 
# Optimization level 1: NumPy 
# 

def biotsavart1(node: np.ndarray, source: np.ndarray) -> np.ndarray: 
    """ Opt level 1: compute magnetic field at a node due to a single source
    """

    a = source[3:6] - source[0:3] 
    b = source[0:3] - node
    c = source[3:6] - node

    cxa = np.cross(c,a) 
    return (mu0*source[6]/(4*pi)) * cxa * (norm2(cxa)**-2) * (np.dot(a,c)/norm2(c)- np.dot(a,b,)/norm(b))


def solve1(nodes: np.ndarray, sources: np.ndarray) -> np.ndarray:
    """ Opt-level 1
    """ 

    Bfield = np.zeros(nodes.shape) 

    for source in sources: 
        for i in range(0, nodes.shape[0]):
            Bfield[i,:] += biotsavart1(nodes[i,:], source)

    return Bfield 


#
# Optimization level 2: Numpy vectorization
#

def normrows(A: np.ndarray) -> np.ndarray: 
    """ opt-level 2: Take the norm of the rows of A 
        Required for numba in opt-level 3
    """
    return np.sqrt(A[:,0]**2 + A[:,1]**2 + A[:,2]**2)


def dotrows(A: np.ndarray, b: np.ndarray): 
    """c = A*b """
    return np.sum(A * b, axis=1)


def biotsavart2(nodes: np.ndarray, source: np.ndarray) -> np.ndarray: 
    """ Vectorized version
    """
    a = source[3:6] - source[0:3] 
    b = np.subtract(source[0:3], nodes)
    c = np.subtract(source[3:6], nodes)

    cxa = np.cross(c, a) 
    coeffs = (mu0*source[6]/(4*pi) / normrows(cxa)**2) * (dotrows(c,a) / normrows(c) - dotrows(b,a) / normrows(b))

    return cxa * coeffs[:,None]


def solve2(nodes, sources): 
    """ Vectorized using NumPy
    """

    Bfield = np.zeros(nodes.shape) 

    for source in sources: 
        Bfield += biotsavart2(nodes, source)

    return Bfield

#
# Optimization level 2: Numpy vectorization (row-major)
# Doesn't seem to be any faster or slower than the default 
#

def normcols(A: np.ndarray) -> np.ndarray: 
    """ opt-level 2: Take the norm of the rows of A 
        Required for numba in opt-level 3
    """
    return np.sqrt(A[0,:]**2 + A[1,:]**2 + A[2,:]**2)


def dotcols(A: np.ndarray, b: np.ndarray): 
    """c = A*b """
    return np.sum(A * np.reshape(b,(3,1)), axis=0)


def biotsavart2_rm(nodes: np.ndarray, source: np.ndarray) -> np.ndarray: 
    """ Vectorized version
    """
    a = source[3:6] - source[0:3] 
    b = -nodes + np.reshape(source[0:3],(3,1)) #np.subtract(np.reshape(source[0:3],(1,3)), nodes)
    c = -nodes + np.reshape(source[3:6],(3,1))

    cxa = np.reshape(np.cross(c, a, axisa=0),(3,nodes.shape[1]))
    coeffs = (mu0*source[6]/(4*pi) / normcols(cxa)**2) * (dotcols(c,a) / normcols(c) - dotcols(b,a) / normcols(b))

    return cxa * np.reshape(coeffs[:,None], (1, nodes.shape[1]))


def solve2_rm(nodes, sources): 
    """ Vectorized using NumPy
    """

    Bfield = np.zeros(nodes.shape) 

    for source in sources: 
        Bfield += biotsavart2_rm(nodes, source)

    return Bfield


# 
# Optimization level 3: Numba 
#  

nbnorm = njit(normrows)
nbdot = njit(dotrows)


@njit
def solve3(nodes: np.ndarray, sources: np.ndarray) -> np.ndarray: 

    Ns = sources.shape[0]
    Bfield = np.zeros(nodes.shape) 

    for i in range(0, Ns): 
        source = sources[i,:]
        a = source[3:6] - source[0:3] 
        b = np.subtract(source[0:3], nodes)
        c = np.subtract(source[3:6], nodes)

        cxa = np.cross(c, a) 
        cxa *= (mu0*source[6]/(4*pi)) * (1 / nbnorm(cxa)**2) * (nbdot(c,a) / nbnorm(c) - nbdot(b,a) / nbnorm(b))
        Bfield = np.add(Bfield, cxa)
    return Bfield


#
# Optimization level 4: parallel Numba
#

@njit
def threadindices(it, Nt, N):

        Nperthread, remainder = divmod(N, Nt)

        if it == 1:
            i1 = 1 
            i2 = i1 + Nperthread - 1 
        elif it == Nt:
            i2 = N 
            i1 = i2 - Nperthread - remainder + 1 
        else:
            i1 = (it-1)*Nperthread + 1
            i2 = i1 + Nperthread - 1 

        return i1,i2

# Íconfig.THREADING_LAYER = 'threadsafe'
@njit(parallel=True)
def solve4(nodes: np.ndarray, sources: np.ndarray):

    Bfield = np.zeros(nodes.shape) 
    Nt = get_num_threads()
    Ns = sources.shape[0]
    Bfield_ = np.zeros((Nt, nodes.shape[0], nodes.shape[1]) )

    for i in prange(0, Nt):

        i1, i2 = threadindices(i, Nt, Ns)

        for j in range(i1, i2+1): #[i1:i2,:]: 
            source = sources[j,:]
            a = source[3:6] - source[0:3] 
            b = np.subtract(source[0:3], nodes)
            c = np.subtract(source[3:6], nodes)

            cxa = np.cross(c, a) # Nx3 matrix
            f = (mu0*source[6]/(4*pi) / nbnorm(cxa)**2) * (nbdot(c,a) / nbnorm(c) - nbdot(b,a) / nbnorm(b))
            cxa[:,0] *= f 
            cxa[:,1] *= f 
            cxa[:,2] *= f
            Bfield_[i] = np.add(Bfield_[i], cxa) 
        
    Bfield = np.sum(Bfield_, axis=0)

    return Bfield

# For convenience:
solvers = [solve0, solve1, solve2, solve3, solve4]